lib/Backends/BackendUtils.cpp

```diff
@@ -81,6 +74,49 @@ runtime::RuntimeBundle::getSymbolInfo(const Named *v) const {
   return it->second;
 }
 
+/// If \p PH is an output placeholder, \returns true.
+/// This is determined by checking if the PH has a user which uses the PH as an
+/// overwritten input.
+bool isOutput(const Placeholder *PH) {
+  // Check if trainable, if it is return true.
+  if (PH->isTraining()) {
+    return true;
+  }
+  for (auto &use : PH->getUsers()) {
+    // Look through the inputs of the PH's users. If an input is overwritten
+    // check if it's the PH, if it is return true.
+    for (unsigned i = 0; i < use.getUser()->getNumInputs(); i++) {
+      if (!use.getUser()->isOverwrittenNthInput(i)) {
+        continue;
+      }
+      auto input = use.getUser()->getNthInput(i);
+      if (input == PH) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// If \p PH is an input placeholder, \returns true.
+bool isInput(const Placeholder *PH) {
+  // Check if trainable, if it is return true.
+  if (PH->isTraining()) {
+    return true;
+  }
+  for (auto &use : PH->getUsers()) {
+    // Check if PH is an input to saveNode
+    if (auto save = llvm::dyn_cast<SaveNode>(use.getUser())) {
```

Comment:
I think there’s still a bug here if a placeholder is the non-overwritten input of a SaveNode; it’ll be erroneously marked as an output.  It’s a degenerate case but it can happen.